src/libclipper/include/clipper/task_executor.hpp

```diff
@@ -162,29 +247,41 @@ class TaskExecutor {
   // containers to the collection when they connect
   std::shared_ptr<ActiveContainers> active_containers_;
   std::unique_ptr<rpc::RPCService> rpc_;
-  Scheduler scheduler_;
   PredictionCache cache_;
   redox::Redox redis_connection_;
   redox::Subscriber redis_subscriber_;
   bool active_ = false;
-  // TODO: Replace the statically configured batch size with dynamic batching
-  const int max_batch_size_ = 5;
   std::mutex inflight_messages_mutex_;
   std::unordered_map<int, std::vector<std::tuple<const long, VersionedModelId,
                                                  std::shared_ptr<Input>>>>
       inflight_messages_;
   std::shared_ptr<metrics::Counter> predictions_counter;
   std::shared_ptr<metrics::Meter> throughput_meter;
   std::shared_ptr<metrics::Histogram> latency_hist;
+  std::unordered_map<const VersionedModelId, ModelQueue, decltype(&versioned_model_hash)> model_queues_;
+  static constexpr int INITIAL_MODEL_QUEUES_MAP_SIZE = 100;
+
+  bool create_model_queue_if_necessary(const VersionedModelId &model_id) {
+    auto queue_added = model_queues_.emplace(std::piecewise_construct,
+                                              std::forward_as_tuple(model_id),
+                                              std::forward_as_tuple());
+    return queue_added.second;
+  }
 
   void on_container_ready(VersionedModelId model_id, int replica_id) {
     std::shared_ptr<ModelContainer> container = active_containers_->get_model_replica(model_id, replica_id);
     if(!container) {
       throw std::runtime_error("TaskExecutor failed to find previously registered active container!");
     }
+    auto model_queue_entry = model_queues_.find(container->model_);
+    if(model_queue_entry == model_queues_.end()) {
+      throw std::runtime_error("Failed to find model queue associated with a previously registered container!");
+    }
     
-    while(true) {
-      auto batch = container->dequeue_predictions(max_batch_size_);
+    for(int i = 0; i < 1000; i++) {
+      Deadline earliest_deadline = model_queue_entry->second.get_earliest_deadline();
```

Comment:
Note that in the current implementation, `get_earliest_deadline()` is blocking. I made a comment up by the function definition to make it non-blocking but repeating it here.